src/java/org/apache/cassandra/service/AsyncProfilerService.java

```diff
@@ -0,0 +1,373 @@
+                                parseDuration(duration),
+                                new File(logDir, validateOutputFileName(outputFileName)));
+
+            String result = maybeInitialize().execute(cmd);
+            logger.debug("Started Async-Profiler: result={}, cmd={}", result, cmd);
+            return true;
+        }
+        catch (IllegalStateException | IllegalArgumentException ex)
+        {
+            throw ex;
+        }
+        catch (Throwable t)
+        {
+            logger.error("Failed to start Async-Profiler", t);
+            return false;
+        }
+    }
+
+    public synchronized boolean stop(String outputFileName)
+    {
+        if (!isRunning())
+            return false;
+
+        try
+        {
+            String cmd = "stop";
+            if (outputFileName != null)
+            {
+                File outputFile = new File(logDir, validateOutputFileName(outputFileName));
+                cmd += ",file=" + outputFile.absolutePath();
+            }
+
+            String result = maybeInitialize().execute(cmd);
+            logger.debug("Stopped Async-Profiler: result={}, cmd={}", result, cmd);
+            return true;
+        }
+        catch (IllegalStateException | IllegalArgumentException e)
+        {
+            throw e;
+        }
+        catch (Throwable e)
+        {
+            logger.error("Failed to stop Async-Profiler", e);
+            return false;
+        }
+    }
+
+    public String execute(String command)
+    {
+        try
+        {
+            String result = maybeInitialize().execute(validateCommand(command));
+            logger.debug("Executed raw command in Async-Profiler: result={}, cmd={}", result, command);
+            return result;
+        }
+        catch (Throwable e)
+        {
+            logger.error("Failed to execute raw Async-Profiler command {}", command, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public List<String> list()
+    {
+        try
+        {
+            createLogDir();
+            return Arrays.stream(new File(logDir).list()).map(File::name).sorted().collect(toList());
+        }
+        catch (Throwable t)
+        {
+            return List.of();
+        }
+    }
+
+    public byte[] fetch(String resultFile)
+    {
+        try
+        {
+            createLogDir();
+            return Files.readAllBytes(new File(logDir, resultFile).toPath());
+        }
+        catch (Throwable t)
+        {
+            return null;
+        }
+    }
+
+    public void purge()
+    {
+        createLogDir();
+        new File(logDir).deleteRecursive();
+    }
+
+    public String status()
+    {
+        try
+        {
+            return maybeInitialize().execute("status");
+        }
+        catch (Throwable t)
+        {
+            return t.getMessage();
+        }
+    }
+
+    public boolean isEnabled()
+    {
+        return profilerInstance != null;
```

Comment:
race condition for the variable access (not looking dangerous in this use case) but from JMM it is not correct..